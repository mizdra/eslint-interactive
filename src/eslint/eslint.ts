export async function shouldUseFlatConfig(): Promise<boolean> {
  try {
    const eslintModule = await import('eslint'); // for eslint v9
    if ('shouldUseFlatConfig' in eslintModule) {
      return await eslintModule.shouldUseFlatConfig();
    }
    const useAtYourOwnRiskModule = await import('eslint/use-at-your-own-risk'); // for eslint v8
    if ('shouldUseFlatConfig' in useAtYourOwnRiskModule) {
      return await useAtYourOwnRiskModule.shouldUseFlatConfig();
    }
  } catch (_e) {
    // noop
  }
  return false; // for eslint v7
}
