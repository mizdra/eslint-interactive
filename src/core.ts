import { join } from 'path';
import { fileURLToPath } from 'url';
import { ESLint } from 'eslint';
import isInstalledGlobally from 'is-installed-globally';
import { format } from './formatter/index.js';
import {
  eslintInteractivePlugin,
  FixRuleOption,
  FixableMaker,
  SuggestionFilter,
  Fix,
  OVERLAPPED_PROBLEM_MESSAGE,
} from './plugin/index.js';
import { unique } from './util/array.js';
import { getCacheDir } from './util/cache.js';
import { filterResultsByRuleId, scanUsedPluginsFromResults } from './util/eslint.js';

const MAX_AUTOFIX_PASSES = 10;

/**
 * Generate results to undo.
 * @param resultsOfLint The results of lint.
 * @returns The results to undo.
 */
function generateResultsToUndo(resultsOfLint: ESLint.LintResult[]): ESLint.LintResult[] {
  return resultsOfLint.map((resultOfLint) => {
    // NOTE: THIS IS HACK.
    return { ...resultOfLint, output: resultOfLint.source };
  });
}

function hasOverlappedProblems(results: ESLint.LintResult[]): boolean {
  return results.flatMap((result) => result.messages).some((message) => message.message === OVERLAPPED_PROBLEM_MESSAGE);
}

/**
 * Get all the rules loaded from eslintrc.
 * @param targetFilePaths The target file paths.
 * @param options The eslint option.
 * @returns The rule ids loaded from eslintrc.
 */
async function getUsedRuleIds(targetFilePaths: string[], options: ESLint.Options): Promise<string[]> {
  const eslintToGetRules = new ESLint(options);
  const configs = await Promise.all(
    targetFilePaths.map(async (filePath) => eslintToGetRules.calculateConfigForFile(filePath)),
  );
  return unique(configs.map((config) => config.rules).flatMap((rules) => Object.keys(rules)));
}

export type Undo = () => Promise<void>;

/** The config of eslint-interactive */
export type Config = {
  patterns: string[];
  rulePaths?: string[] | undefined;
  extensions?: string[] | undefined;
  formatterName?: string;
  cache?: boolean;
  cacheLocation?: string;
  cwd?: string;
};

/** Default config of `Core` */
export const DEFAULT_BASE_CONFIG = {
  cache: true,
  cacheLocation: join(getCacheDir(), '.eslintcache'),
  formatterName: 'codeframe',
};

/**
 * The core of eslint-interactive.
 * It uses ESLint's Node.js API to output a summary of problems, fix problems, apply suggestions, etc.
 */
export class Core {
  readonly config: Config;
  /** The base options of ESLint */
  readonly baseOptions: ESLint.Options;

  constructor(config: Config) {
    this.config = config;
    this.baseOptions = {
      cache: this.config.cache ?? DEFAULT_BASE_CONFIG.cache,
      cacheLocation: this.config.cacheLocation ?? DEFAULT_BASE_CONFIG.cacheLocation,
      rulePaths: this.config.rulePaths,
      extensions: this.config.extensions,
      cwd: this.config.cwd,
    };
  }

  /**
   * Lint project.
   * @returns The results of linting
   */
  async lint(): Promise<ESLint.LintResult[]> {
    const eslint = new ESLint(this.baseOptions);
    const results = await eslint.lintFiles(this.config.patterns);
    return results;
  }

  /**
   * Returns summary of lint results.
   * @param results The lint results of the project to print summary
   */
  formatResultSummary(results: ESLint.LintResult[]): string {
    // get used plugins from `results`
    const plugins = scanUsedPluginsFromResults(results);

    // get `rulesMeta` from `results`
    const eslint = new ESLint({
      ...this.baseOptions,
      overrideConfig: { plugins },
    });
    // NOTE: `getRulesMetaForResults` is a feature added in ESLint 7.29.0.
    // Therefore, the function may not exist in versions lower than 7.29.0.
    const rulesMeta: ESLint.LintResultData['rulesMeta'] = eslint.getRulesMetaForResults?.(results) ?? {};

    return format(results, { rulesMeta: rulesMeta, cwd: this.config.cwd ?? process.cwd() });
  }

  /**
   * Returns details of lint results.
   * @param results The lint results of the project to print summary
   * @param ruleIds The rule ids to print details
   */
  async formatResultDetails(results: ESLint.LintResult[], ruleIds: (string | null)[]): Promise<string> {
    const eslint = new ESLint(this.baseOptions);
    const formatterName = this.config.formatterName ?? DEFAULT_BASE_CONFIG.formatterName;

    // When eslint-interactive is installed globally, eslint-formatter-codeframe will also be installed globally.
    // On the other hand, `eslint.loadFormatter` cannot load the globally installed formatter by name. So here it loads them by path.
    const resolvedFormatterNameOrPath =
      isInstalledGlobally && formatterName === 'codeframe'
        ? fileURLToPath(
            // @ts-expect-error
            await import.meta.resolve(
              'eslint-formatter-codeframe',
              // @ts-expect-error
              await import.meta.resolve('eslint-interactive'),
            ),
          )
        : formatterName;

    const formatter = await eslint.loadFormatter(resolvedFormatterNameOrPath);
    return formatter.format(filterResultsByRuleId(results, ruleIds));
  }

  /**
   * Run `eslint --fix`.
   * @param ruleIds The rule ids to fix
   */
  async applyAutoFixes(results: ESLint.LintResult[], ruleIds: string[]): Promise<Undo> {
    return await this.fix(results, ruleIds, { name: 'applyAutoFixes', args: {} });
  }

  /**
   * Add disable comments per line.
   * @param results The lint results of the project to add disable comments
   * @param ruleIds The rule ids to add disable comments
   * @param description The description of the disable comments
   */
  async disablePerLine(results: ESLint.LintResult[], ruleIds: string[], description?: string): Promise<Undo> {
    return await this.fix(results, ruleIds, { name: 'disablePerLine', args: { description } });
  }

  /**
   * Add disable comments per file.
   * @param results The lint results of the project to add disable comments
   * @param ruleIds The rule ids to add disable comments
   * @param description The description of the disable comments
   */
  async disablePerFile(results: ESLint.LintResult[], ruleIds: string[], description?: string): Promise<Undo> {
    return await this.fix(results, ruleIds, { name: 'disablePerFile', args: { description } });
  }

  /**
   * Apply suggestions.
   * @param results The lint results of the project to apply suggestions
   * @param ruleIds The rule ids to apply suggestions
   * @param filter The script to filter suggestions
   * */
  async applySuggestions(results: ESLint.LintResult[], ruleIds: string[], filter: SuggestionFilter): Promise<Undo> {
    return await this.fix(results, ruleIds, { name: 'applySuggestions', args: { filter } });
  }

  /**
   * Make forcibly fixable and run `eslint --fix`.
   * @param results The lint results of the project to apply suggestions
   * @param ruleIds The rule ids to apply suggestions
   * @param fixableMaker The function to make `Linter.LintMessage` forcibly fixable.
   * */
  async makeFixableAndFix(results: ESLint.LintResult[], ruleIds: string[], fixableMaker: FixableMaker): Promise<Undo> {
    return await this.fix(results, ruleIds, { name: 'makeFixableAndFix', args: { fixableMaker } });
  }

  /**
   * Fix source codes.
   * @param fix The fix information to do.
   */
  private async fix(resultsOfLint: ESLint.LintResult[], ruleIds: string[], fix: Fix): Promise<Undo> {
    // NOTE: Extract only necessary results and files for performance
    const filteredResultsOfLint = filterResultsByRuleId(resultsOfLint, ruleIds);
    const targetFilePaths = filteredResultsOfLint.map((result) => result.filePath);
    const usedRuleIds = await getUsedRuleIds(targetFilePaths, this.baseOptions);

    // TODO: refactor
    let results = filteredResultsOfLint;
    for (let i = 0; i < MAX_AUTOFIX_PASSES; i++) {
      const eslint = new ESLint({
        ...this.baseOptions,
        // This is super hack to load ESM plugin/rule.
        // ref: https://github.com/eslint/eslint/issues/15453#issuecomment-1001200953
        plugins: {
          'eslint-interactive': eslintInteractivePlugin,
        },
        overrideConfig: {
          plugins: ['eslint-interactive'],
          rules: {
            'eslint-interactive/fix': [2, { results, ruleIds, fix } as FixRuleOption],
            // Turn off all rules except `eslint-interactive/fix` when fixing for performance.
            ...Object.fromEntries(usedRuleIds.map((ruleId) => [ruleId, 'off'])),
          },
        },
        // NOTE: Only fix the `fix` rule problems.
        fix: (message) => message.ruleId === 'eslint-interactive/fix',
        // Don't interpret lintFiles arguments as glob patterns for performance.
        globInputPaths: false,
      });
      const resultsToFix = await eslint.lintFiles(targetFilePaths);
      await ESLint.outputFixes(resultsToFix);
      if (!hasOverlappedProblems(resultsToFix)) break;
      results = await this.lint();
    }

    return async () => {
      const resultsToUndo = generateResultsToUndo(filteredResultsOfLint);
      await ESLint.outputFixes(resultsToUndo);
    };
  }
}
