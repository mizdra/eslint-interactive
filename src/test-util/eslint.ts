import { Rule, Linter, ESLint } from 'eslint';

export function fakeFix(arg?: Partial<Rule.Fix>): Rule.Fix {
  return {
    range: [0, 0],
    text: '',
    ...arg,
  };
}

export function fakeLintMessage(arg?: Partial<Linter.LintMessage>): Linter.LintMessage {
  return {
    column: 0,
    line: 0,
    ruleId: 'ruleId',
    message: 'message',
    severity: 2,
    ...arg,
  };
}

export function fakeLintResult(arg?: Partial<ESLint.LintResult>): ESLint.LintResult {
  return {
    filePath: '',
    messages: [],
    errorCount: 0,
    fatalErrorCount: 0,
    warningCount: 0,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
    usedDeprecatedRules: [],
    suppressedMessages: [],
    ...arg,
  };
}

function fakeSuggestion(arg?: Partial<Linter.LintSuggestion>): Linter.LintSuggestion {
  return {
    desc: 'description',
    fix: fakeFix(),
    ...arg,
  };
}

export function fakeSuggestions(): Linter.LintSuggestion[] {
  return [fakeSuggestion(), fakeSuggestion()];
}
