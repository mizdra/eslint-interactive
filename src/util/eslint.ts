import { AST, ESLint } from 'eslint';
import type { Comment } from 'estree';
import { unique } from './array.js';
import { notEmpty } from './type-check.js';

const COMMENT_RE =
  /^\s*(?<header>eslint-disable|eslint-disable-next-line)\s+(?<ruleList>[@a-z0-9\-_$/]*(?:\s*,\s*[@a-z0-9\-_$/]*)*(?:\s*,)?)(?:\s+--\s+(?<description>.*\S))?\s*$/u;

const SHEBANG_PATTERN = /^#!.+?\r?\n/u;

/** `results` 内で使われているプラグインの名前のリストを洗い出して返す */
export function scanUsedPluginsFromResults(results: ESLint.LintResult[]): string[] {
  const plugins = results
    .flatMap((result) => result.messages) // messages: Linter.LintMessage[]
    .map((message) => message.ruleId) // ruleIds: (string | undefined)[]
    .filter(notEmpty) // ruleIds: string[]
    .map((ruleId) => {
      const parts = ruleId.split('/');
      if (parts.length === 1) return undefined; // ex: 'rule-a'
      if (parts.length === 2) return parts[0]; // ex: 'plugin/rule-a'
      if (parts.length === 3) return `${parts[0]}/${parts[1]}`; // ex: '@scoped/plugin/rule-a'
      return undefined; // invalid ruleId
    }) // plugins: string[]
    .filter(notEmpty);
  return unique(plugins);
}

export type DisableComment = {
  type: 'Block' | 'Line';
  scope: 'next-line' | 'file';
  ruleIds: string[];
  description?: string;
  range: [number, number];
};

/**
 * コメントを ESLint の disable comment としてパースする。
 * disable comment としてパースできなかった場合は undefined を返す。
 *
 * ## 参考: disable comment の構造
 * /* eslint-disable-next-line rule-a, rule-b, rule-c, rule-d -- I'm the rules.
 *    ^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^ ^^^^^^^^^^^^^^
 *    |                        |                              |  |
 *    header                   |                              |  |
 *                             ruleList                       |  |
 *                                            descriptionHeader  |
 *                                                               description
 */
export function parseDisableComment(comment: Comment): DisableComment | undefined {
  // NOTE: コメントノードには必ず range があるはずだが、型上は optional なので、
  // range がない場合はパースに失敗した扱いにする。
  if (!comment.range) return undefined;

  const result = COMMENT_RE.exec(comment.value);
  if (!result) return undefined;
  if (!result.groups) return undefined;

  const { header, ruleList, description } = result.groups;
  const ruleIds = ruleList
    .split(',')
    .map((r) => r.trim())
    // 空文字は除外しておく
    .filter((ruleId) => ruleId !== '');

  const scope = header === 'eslint-disable-next-line' ? 'next-line' : 'file';
  // file scope comment must be block-style.
  if (scope === 'file' && comment.type === 'Line') return undefined;

  return {
    type: comment.type,
    scope: header === 'eslint-disable-next-line' ? 'next-line' : 'file',
    ruleIds: ruleIds,
    // description is optional
    ...(description === '' || description === undefined ? {} : { description }),
    range: comment.range,
  };
}

/**
 * Convert `DisableComment` to comment text.
 */
export function toCommentText({ type, scope, ruleIds, description }: Omit<DisableComment, 'range'>): string {
  const header = scope === 'next-line' ? 'eslint-disable-next-line' : 'eslint-disable';
  const ruleList = unique(ruleIds).join(', ');
  if (type === 'Line') {
    if (description === undefined) {
      return `// ${header} ${ruleList}`;
    } else {
      return `// ${header} ${ruleList} -- ${description}`;
    }
  } else {
    if (description === undefined) {
      return `/* ${header} ${ruleList} */`;
    } else {
      return `/* ${header} ${ruleList} -- ${description} */`;
    }
  }
}

/**
 * Create the results with only messages with the specified rule ids.
 * @param results The lint results.
 * @param ruleIds The rule ids.
 * @returns The results with only messages with the specified rule ids
 */
export function filterResultsByRuleId(results: ESLint.LintResult[], ruleIds: (string | null)[]): ESLint.LintResult[] {
  return results.map((result) => {
    return {
      ...result,
      messages: result.messages.filter((message) => ruleIds.includes(message.ruleId)),
    };
  });
}

/**
 * push rule ids to the disable comment and return the new comment node.
 * @param comment The comment node to be modified
 * @param ruleIds The rule ids to be added
 * @returns The new comment node
 */
export function pushRuleIdsToDisableComment(comment: DisableComment, ruleIds: string[]): DisableComment {
  return {
    ...comment,
    ruleIds: unique([...comment.ruleIds, ...ruleIds]),
  };
}

/**
 * Merge the ruleIds and description of the disable comments.
 * @param a The ruleIds and description of first disable comment
 * @param b The ruleIds and description of second disable comment
 * @returns The ruleIds and description of merged disable comment
 */
export function mergeRuleIdsAndDescription(
  a: { ruleIds: string[]; description?: string },
  b: { ruleIds: string[]; description?: string },
): { ruleIds: string[]; description?: string } {
  const ruleIds = unique([...a.ruleIds, ...b.ruleIds]);
  const description =
    a.description !== undefined && b.description !== undefined
      ? `${a.description}, ${b.description}`
      : a.description !== undefined && b.description === undefined
      ? a.description
      : a.description === undefined && b.description !== undefined
      ? b.description
      : undefined;
  if (description === undefined) return { ruleIds };
  return { ruleIds, description };
}

/**
 * Find shebang from the first line of the file.
 * @param sourceCodeText The source code text of the file.
 * @returns The information of shebang. If the file does not have shebang, return null.
 */
export function findShebang(sourceCodeText: string): { range: AST.Range } | null {
  const result = SHEBANG_PATTERN.exec(sourceCodeText);
  if (!result) return null;
  return { range: [0, result[0].length] };
}
