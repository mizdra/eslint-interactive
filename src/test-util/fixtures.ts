import { exec } from 'node:child_process';
import { randomUUID } from 'node:crypto';
import { rm } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { defineIFFCreator } from '@mizdra/inline-fixture-files';
import fse from 'fs-extra/esm';

const execPromise = promisify(exec);

const rootDir = join(dirname(fileURLToPath(import.meta.url)), '..', '..');

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const fixtureDir = join(rootDir, 'tmp/fixture', process.env['VITEST_POOL_ID']!);
export const createIFF = defineIFFCreator({ generateRootDir: () => join(fixtureDir, randomUUID()) });

/**
 * Returns a string containing the stitched together contents of the file modified by fix.
 * To make the snapshot easier to read, the name of the file is inserted at the beginning of the contents of each file.
 */
export async function getSnapshotOfChangedFiles(): Promise<string> {
  const { stdout } = await execPromise(`diff -qr fixtures fixtures-tmp | cut -d " " -f 4 | xargs tail -n +1`, {
    cwd: rootDir,
  });
  return stdout.toString();
}

export async function setupFixturesCopy() {
  await rm(resolve(rootDir, 'fixtures-tmp'), { recursive: true, force: true });
  await fse.copy(resolve(rootDir, 'fixtures'), resolve(rootDir, 'fixtures-tmp'));
}

export async function cleanupFixturesCopy() {
  await rm(resolve(rootDir, 'fixtures-tmp'), { recursive: true, force: true });
}
