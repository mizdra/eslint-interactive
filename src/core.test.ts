import { constants, cp, readFile } from 'node:fs/promises';
import { dirname, join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';
import dedent from 'dedent';
import { ESLint, Linter } from 'eslint';
import { resolve } from 'import-meta-resolve';
import { beforeEach, describe, expect, test, vi } from 'vitest';
import { Core } from './core.js';
import { createIFF } from './test-util/fixtures.js';

const testIf = (condition: boolean) => (condition ? test : test.skip);

const rootDir = join(dirname(fileURLToPath(import.meta.url)), '..');

// Normalize `message` for the snapshot.
function normalizeMessage(message: Linter.LintMessage) {
  return {
    ruleId: message.ruleId,
    severity: message.severity,
  } satisfies Partial<Linter.LintMessage>;
}

// Normalize `results` for the snapshot.
function normalizeResults(results: ESLint.LintResult[], fixtureDir: string) {
  return results.map((result) => {
    // Usually, `filePath` changes depending on the environment, and the snapshot will fail.
    // So, remove the current directory from `filePath`.
    let filePath = result.filePath
      // for windows
      .replace(/\\/gu, '/');
    if (fixtureDir)
      filePath = filePath.replace(fixtureDir, '<fixture>').replace(relative(process.cwd(), fixtureDir), '<fixture>');
    return {
      filePath,
      errorCount: result.errorCount,
      warningCount: result.warningCount,
      messages: result.messages.map(normalizeMessage),
    };
  });
}

function countWarnings(results: ESLint.LintResult[]): number {
  return results.map((result) => result.warningCount).reduce((acc, num) => acc + num, 0);
}

const iff = await createIFF({
  'src/prefer-const.js': dedent`
    let a = 1;
  `,
  'src/arrow-body-style.js': dedent`
    () => (
      () => (
        () => (
          () => (
            () => (
              () => (
                () => (
                  () => (
                    () => (
                      () => (
                        () => (
                          () => (
                            0
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    );
  `,
  'src/import-order.js': dedent`
    import b from 'b';
    import a from 'a';
  `,
  'src/ban-exponentiation-operator.js': dedent`
    2 ** 2;
  `,
  'src/no-unused-vars.js': dedent`
    const a = 1;
  `,
  'src/warn.js': dedent`
    let a = 1;
  `,
  '.eslintrc.js': dedent`
    module.exports = {
      root: true,
      parserOptions: {
        ecmaVersion: 2022,
        sourceType: 'module',
      },
      overrides: [
        { files: ['prefer-const.js'], rules: { 'prefer-const': 'error' } },
        { files: ['arrow-body-style.js'], rules: { 'arrow-body-style': ['error', 'always'] } },
        { files: ['import-order.js'], rules: { 'import/order': 'error' } },
        { files: ['ban-exponentiation-operator.js'], rules: { 'ban-exponentiation-operator': 'error' } },
        { files: ['no-unused-vars.js'], rules: { 'no-unused-vars': 'error' } },
        { files: ['warn.js'], rules: { 'prefer-const': 'warn' } },
      ],
    };
  `,
  'package.json': '{ "type": "commonjs" }',
  'rules': async (path) =>
    cp(join(rootDir, 'fixtures/rules'), path, { mode: constants.COPYFILE_FICLONE, recursive: true }),
});

const core = new Core({
  patterns: ['src'],
  // For some reason, the test fails if `formatterName === 'codeframe'`.
  // So here we overwrite it.
  formatterName: fileURLToPath(resolve('eslint-formatter-codeframe', import.meta.url)),
  cwd: iff.rootDir,
  eslintOptions: { type: 'eslintrc', rulePaths: ['rules'] },
});

beforeEach(async () => {
  await iff.reset();
});

describe('Core', () => {
  describe('constructor', () => {
    test('pass options to eslint', async () => {
      const iff = await createIFF({
        'src/index.js': 'let a = 1;',
        'src/index.mjs': '2 ** 2;',
        'rules': async (path) =>
          cp(join(rootDir, 'fixtures/rules'), path, { mode: constants.COPYFILE_FICLONE, recursive: true }),
        'package.json': '{ "type": "commonjs" }',
      });
      const core = new Core({
        patterns: ['src'],
        cwd: iff.rootDir,
        eslintOptions: {
          type: 'eslintrc',
          useEslintrc: false,
          rulePaths: ['rules'],
          extensions: ['.js', '.mjs'],
          overrideConfig: {
            parserOptions: {
              ecmaVersion: 2022,
              sourceType: 'module',
            },
            rules: { 'prefer-const': 'error', 'ban-exponentiation-operator': 'error' },
          },
        },
      });
      const results = await core.lint();
      expect(normalizeResults(results, iff.rootDir)).toMatchSnapshot();
    });
  });
  describe('lint', () => {
    test('returns lint results', async () => {
      const results = await core.lint();
      expect(normalizeResults(results, iff.rootDir)).toMatchSnapshot();
    });
    test('filters warnings with --quiet option', async () => {
      const coreWithoutQuiet = new Core({
        ...core.config,
        quiet: false,
      });
      const resultsWithoutQuiet = await coreWithoutQuiet.lint();
      expect(countWarnings(resultsWithoutQuiet)).not.toEqual(0);

      const coreWithQuiet = new Core({
        ...core.config,
        quiet: true,
      });
      const resultsWithQuiet = await coreWithQuiet.lint();
      expect(countWarnings(resultsWithQuiet)).toEqual(0);
    });
  });
  // This test fails because the documentation url format is not supported in eslint 7.x.x and 8.0.0. Therefore, ignore this test.
  testIf(!ESLint.version.startsWith('7.'))('printSummaryOfResults', async () => {
    const results = await core.lint();
    vi.spyOn(ESLint.prototype, 'getRulesMetaForResults').mockImplementationOnce(() => {
      return {
        'prefer-const': {
          docs: {
            url: 'https://example.com',
          },
        },
      };
    });
    expect(core.formatResultSummary(results)).toMatchSnapshot();
  });
  test('printDetailsOfResults', async () => {
    const results = await core.lint();
    expect(
      (await core.formatResultDetails(results, ['import/order', 'ban-exponentiation-operator']))
        // for windows
        .replace(/\\/gu, '/')
        .replaceAll(iff.rootDir, '<fixture>')
        .replaceAll(relative(process.cwd(), iff.rootDir), '<fixture>'),
    ).toMatchSnapshot();
  });
  describe('applyAutoFixes', () => {
    test('basic', async () => {
      const results = await core.lint();
      const original = await readFile(iff.paths['src/prefer-const.js'], 'utf-8');

      const undo = await core.applyAutoFixes(results, ['prefer-const']);

      expect(await readFile(iff.paths['src/prefer-const.js'], 'utf-8')).toMatchSnapshot();
      await undo();
      expect(await readFile(iff.paths['src/prefer-const.js'], 'utf-8')).toEqual(original);
    });
    test('fix overlapped problems', async () => {
      // NOTE: It can fix up to 11 overlapping errors. This is due to a constraints imposed by ESLint to prevent infinite loops.
      // ref: https://github.com/eslint/eslint/blob/5d60812d440762dff72420714273c714c4c5d074/lib/linter/linter.js#L44
      const results = await core.lint();
      const original = await readFile(iff.paths['src/arrow-body-style.js'], 'utf-8');

      const undo = await core.applyAutoFixes(results, ['arrow-body-style']);

      expect(await readFile(iff.paths['src/arrow-body-style.js'], 'utf-8')).toMatchSnapshot();
      await undo();
      expect(await readFile(iff.paths['src/arrow-body-style.js'], 'utf-8')).toEqual(original);
    });
  });
  test('disablePerLine', async () => {
    const results = await core.lint();
    const original = await readFile(iff.paths['src/prefer-const.js'], 'utf-8');

    const undo = await core.disablePerLine(results, ['prefer-const']);

    expect(await readFile(iff.paths['src/prefer-const.js'], 'utf-8')).toMatchSnapshot();
    await undo();
    expect(await readFile(iff.paths['src/prefer-const.js'], 'utf-8')).toEqual(original);
  });
  test('disablePerFile', async () => {
    const results = await core.lint();
    const original = await readFile(iff.paths['src/prefer-const.js'], 'utf-8');

    const undo = await core.disablePerFile(results, ['prefer-const']);

    expect(await readFile(iff.paths['src/prefer-const.js'], 'utf-8')).toMatchSnapshot();
    await undo();
    expect(await readFile(iff.paths['src/prefer-const.js'], 'utf-8')).toEqual(original);
  });
  test('applySuggestions', async () => {
    const results = await core.lint();
    const original = await readFile(iff.paths['src/no-unused-vars.js'], 'utf-8');

    const undo = await core.applySuggestions(results, ['no-unsafe-negation'], (suggestions) => suggestions[0]);

    expect(await readFile(iff.paths['src/no-unused-vars.js'], 'utf-8')).toMatchSnapshot();
    await undo();
    expect(await readFile(iff.paths['src/no-unused-vars.js'], 'utf-8')).toEqual(original);
  });
  test('makeFixableAndFix', async () => {
    const results = await core.lint();
    const original = await readFile(iff.paths['src/no-unused-vars.js'], 'utf-8');

    const undo = await core.makeFixableAndFix(results, ['no-unused-vars'], (_message, node) => {
      if (!node || !node.range) return null;
      return { range: [node.range[0], node.range[0]], text: '_' };
    });

    expect(await readFile(iff.paths['src/no-unused-vars.js'], 'utf-8')).toMatchSnapshot();
    await undo();
    expect(await readFile(iff.paths['src/no-unused-vars.js'], 'utf-8')).toEqual(original);
  });
});
