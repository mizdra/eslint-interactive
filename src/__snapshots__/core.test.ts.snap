// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Core > applyAutoFixes > basic 1`] = `"const a = 1;"`;

exports[`Core > applyAutoFixes > fix overlapped problems 1`] = `
"() => {return (
  () => {return (
    () => {return (
      () => {return (
        () => {return (
          () => {return (
            () => {return (
              () => {return (
                () => {return (
                  () => {return (
                    () => (
                      () => (
                        0
                      )
                    )
                  )}
                )}
              )}
            )}
          )}
        )}
      )}
    )}
  )}
)};"
`;

exports[`Core > applySuggestions 1`] = `"const a = 1;"`;

exports[`Core > constructor > pass options to eslint 1`] = `
[
  {
    "errorCount": 1,
    "filePath": "<fixture>/src/index.js",
    "messages": [
      {
        "ruleId": "prefer-const",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "filePath": "<fixture>/src/index.mjs",
    "messages": [
      {
        "ruleId": "ban-exponentiation-operator",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
]
`;

exports[`Core > disablePerFile 1`] = `
"/* eslint-disable ban-exponentiation-operator */
2 ** 2;"
`;

exports[`Core > disablePerLine 1`] = `
"// eslint-disable-next-line ban-exponentiation-operator
2 ** 2;"
`;

exports[`Core > lint > returns lint results 1`] = `
[
  {
    "errorCount": 12,
    "filePath": "<fixture>/src/arrow-body-style.js",
    "messages": [
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
      {
        "ruleId": "arrow-body-style",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "filePath": "<fixture>/src/ban-exponentiation-operator.js",
    "messages": [
      {
        "ruleId": "ban-exponentiation-operator",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "filePath": "<fixture>/src/import-order.js",
    "messages": [
      {
        "ruleId": "import/order",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "filePath": "<fixture>/src/no-unused-vars.js",
    "messages": [
      {
        "ruleId": "no-unused-vars",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "filePath": "<fixture>/src/prefer-const.js",
    "messages": [
      {
        "ruleId": "prefer-const",
        "severity": 2,
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 0,
    "filePath": "<fixture>/src/warn.js",
    "messages": [
      {
        "ruleId": "prefer-const",
        "severity": 1,
      },
    ],
    "warningCount": 1,
  },
]
`;

exports[`Core > makeFixableAndFix 1`] = `"const _a = 1;"`;

exports[`Core > printDetailsOfResults 1`] = `
"[31merror[39m: [1mBan exponentiation operator[22m [2m(ban-exponentiation-operator)[22m at [32m<fixture>/src/ban-exponentiation-operator.js:1:1[39m:
> 1 | 2 ** 2;
    | ^


[31merror[39m: [1mDefinition for rule 'import/order' was not found[22m [2m(import/order)[22m at [32m<fixture>/src/import-order.js:1:1[39m:
> 1 | import b from 'b';
    | ^
  2 | import a from 'a';


[31m[1m2 errors found.[22m[39m"
`;

exports[`Core > printSummaryOfResults 1`] = `
"[1m- 6 files (0 file passed, [91m6 files failed[39m) checked.[22m
[1m- 17 problems ([31m16 errors[39m, [33m1 warning[39m) found.[22m
╔═════════════════════════════╤═══════╤═════════╤════════════╤═════════════════╗
║ Rule                        │ Error │ Warning │ is fixable │ has suggestions ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ arrow-body-style            │ [31m[1m12[39m[22m    │ 0       │ [31m[1m12[39m[22m         │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ ban-exponentiation-operator │ [31m[1m1[39m[22m     │ 0       │ 0          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ import/order                │ [31m[1m1[39m[22m     │ 0       │ 0          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ no-unused-vars              │ [31m[1m1[39m[22m     │ 0       │ 0          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ ]8;;https://example.comprefer-const]8;;                │ [31m[1m1[39m[22m     │ [31m[1m1[39m[22m       │ [31m[1m2[39m[22m          │ 0               ║
╚═════════════════════════════╧═══════╧═════════╧════════════╧═════════════════╝
"
`;
